drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,396 @@
+
+#if defined(CONFIG_SOC_SERIES_GD32F4XX)
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x
+
+static const uint16_t acq_time_tbl[8] = {3, 15, 28, 56, 84, 112, 144, 480};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(3),
+	SMP_TIME(15),
+	SMP_TIME(28),
+	SMP_TIME(56),
+	SMP_TIME(84),
+	SMP_TIME(112),
+	SMP_TIME(144),
+	SMP_TIME(480)
+};
+#else
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x##POINT5
+
+static const uint16_t acq_time_tbl[8] = {2, 8, 14, 29, 42, 56, 72, 240};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(1),
+	SMP_TIME(7),
+	SMP_TIME(13),
+	SMP_TIME(28),
+	SMP_TIME(41),
+	SMP_TIME(55),
+	SMP_TIME(71),
+	SMP_TIME(239)
+};
+#endif
+
+struct adc_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	uint8_t channels;
+	const struct pinctrl_dev_config *pcfg;
+	void (*irq_config_func)(void);
+};
+
+struct adc_gd32_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+};
+
+/* Use a flag to protect the shared irq handler initialization. */
+static bool init_irq = true;
+
+static void adc_gd32_isr(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
+		*data->buffer++ = ADC_RDATA(cfg->reg);
+
+		/* Disable EOC interrupt. */
+		ADC_CTL0(cfg->reg) &= ~ADC_CTL0_EOCIE;
+		/* Clear EOC bit. */
+		ADC_STAT(cfg->reg) &= ~ADC_STAT_EOC;
+
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
+	const struct device *dev = data->dev;
+	const struct adc_gd32_config *cfg = dev->config;
+
+	data->repeat_buffer = data->buffer;
+
+	/* Enable EOC interrupt */
+	ADC_CTL0(cfg->reg) |= ADC_CTL0_EOCIE;
+
+	/* Set ADC software conversion trigger. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_SWRCST;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static inline void adc_gd32_calibration(const struct adc_gd32_config *cfg)
+{
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_RSTCLB;
+	/* Wait for calibration registers initialized. */
+	while (ADC_CTL1(cfg->reg) & ADC_CTL1_RSTCLB) {
+	}
+
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_CLB;
+	/* Wait for calibration complete. */
+	while (ADC_CTL1(cfg->reg) & ADC_CTL1_CLB) {
+	}
+}
+
+static int adc_gd32_configure_sampt(uint8_t channel, uint16_t acq_time)
+{
+	uint8_t index, offset;
+
+	/* Acquisitioin time unit is adc clock cycle. */
+	if (ADC_ACQ_TIME_UNIT(acq_time) != ADC_ACQ_TIME_TICKS) {
+		return -EINVAL;
+	}
+
+	index = 0;
+
+	if (acq_time != ADC_ACQ_TIME_DEFAULT) {
+		for (uint8_t i = 0; i < ARRAY_SIZE(acq_time_tbl); i++) {
+			if (ADC_ACQ_TIME_VALUE(acq_time) <= acq_time_tbl[i]) {
+				index = i;
```

Comment:
Seems like you should be able to use index as the incrementor instead of creating another `uint8_t i`?

Reply:
done.